### ISSUE TITLE
Add AsyncWaitersCount property to SemaphoreSlim

### ISSUE THREAD

**Issue by omariom** [dotnet/runtime#5234]
`SemaphoreSlim` is quite powerful construct for async computations. 
But, IMO, it lacks an important property - `AsyncWaitersCount` - which could report the number of waiters blocked on `WaitAsync` method. 

With this property we could check if there are too many waiters queued already 
so that it is guaranteed (given the size of the queue and average time waiters spend in it) that any next waiter will miss its SLA threshold. Increasing the waiters queue in such case will only make the situation worse.

Basically I want to be able to do this:

``` C#

if (semaphoreSlim.AsyncWaitersCount > 100)
    throw new ServerIsBusyException();

await semaphoreSlim.WaitAsync();
...
```
